File: backend/app/mathematics/difficulty.py
from __future__ import annotations


def select_tier(current_tier: int, skill_score: float, parent_override: int | None = None) -> int:
    if parent_override is not None:
        return max(1, min(6, parent_override))
    if skill_score >= 0.90:
        return min(current_tier + 1, 6)
    if skill_score < 0.60:
        return max(current_tier - 1, 1)
    return current_tier

File: frontend/src/engine/math/difficulty.ts
import type { Tier, TierSelectionInput } from './types';

function clampTier(value: number): Tier {
  return Math.max(1, Math.min(6, value)) as Tier;
}

export function selectTier(input: TierSelectionInput): Tier {
  const { currentTier, skillScore, parentOverride } = input;
  if (parentOverride !== undefined && parentOverride !== null) {
    return clampTier(parentOverride);
  }
  if (skillScore >= 0.9) {
    return clampTier(currentTier + 1);
  }
  if (skillScore < 0.6) {
    return clampTier(currentTier - 1);
  }
  return currentTier;
}
